### ISSUE TITLE
Incorrect results triggered by query partitioning

### ISSUE THREAD

**Issue by Lincyaw** [FalkorDB/FalkorDB#37]
Version: Docker image falkordb/falkordb:edge
OS: Ubuntu 22.04
API/Driver: Cypher

According to the following query, the 18 should equals to 10+12

```
lab:6380> GRAPH.QUERY 0 "MATCH (n1:L3)-[]-(n2:L2)-[]-(n3:L1:L2)-[r1:T1]-(n4:L3:L0:L4) RETURN COUNT(*)"
1) 1) "COUNT(*)"
2) 1) 1) (integer) 18
3) 1) "Cached execution: 0"
   2) "Query internal execution time: 1.820675 milliseconds"
lab:6380> GRAPH.QUERY 0 "MATCH (n1:L3)<-[]-(n2:L2)-[]-(n3:L1:L2)-[r1:T1]-(n4:L3:L0:L4) RETURN COUNT(*)"
1) 1) "COUNT(*)"
2) 1) 1) (integer) 10
3) 1) "Cached execution: 0"
   2) "Query internal execution time: 0.738668 milliseconds"
lab:6380> GRAPH.QUERY 0 "MATCH (n1:L3)-[]->(n2:L2)-[]-(n3:L1:L2)-[r1:T1]-(n4:L3:L0:L4) RETURN COUNT(*)"
1) 1) "COUNT(*)"
2) 1) 1) (integer) 12
3) 1) "Cached execution: 0"
   2) "Query internal execution time: 0.759969 milliseconds"
lab:6380> GRAPH.QUERY 0 "MATCH (n)-[r]-(n) RETURN n,r LIMIT 10"
1) 1) "n"
   2) "r"
2) (empty array)
3) 1) "Cached execution: 0"
   2) "Query internal execution time: 0.584519 milliseconds"
```

The create statement is attached below for your reproduction
[1692914926.9295375.log](https://github.com/FalkorDB/FalkorDB/files/12434890/1692914926.9295375.log)

**Comment by gkorland** [comment]
@Lincyaw you're right the semantic is not well defined in Cypher, counting different overlapping paths on the first example is considered as a single result for the count.

So in your example, two directed paths that have the same components will be count as a single path.

You can overcome this by counting unique paths.

```
GRAPH.QUERY 0 "MATCH p=(n1:L3)-[]-(n2:L2)-[]-(n3:L1:L2)-[r1:T1]-(n4:L3:L0:L4) RETURN COUNT(p)"
```

**Comment by Lincyaw** [comment]
<img width="1536" alt="image" src="https://github.com/FalkorDB/FalkorDB/assets/53161583/608e1694-8d66-4711-bc3b-5d6e0148bee7">


You are right, but for the same queries, the neo4j's results seems different from the FalkerDB. Is that normal?

**Comment by gkorland** [comment]
Yes, the semantic is not well defined in Cypher so you might find different vendors interpenetration

**Comment by Lincyaw** [comment]
After searching the Cypher Specification, I found the following use cases. 

https://github.com/opencypher/openCypher/blob/master/tck/features/clauses/match/Match8.feature

And in the following test case

```
 Scenario: [2] Counting rows after MATCH, MERGE, OPTIONAL MATCH
    Given an empty graph
    And having executed:
      """
      CREATE (a:A), (b:B)
      CREATE (a)-[:T1]->(b),
             (b)-[:T2]->(a)
      """
    When executing query:
      """
      MATCH (a)
      MERGE (b)
      WITH *
      OPTIONAL MATCH (a)--(b)
      RETURN count(*)
      """
    Then the result should be, in any order:
      | count(*) |
      | 6        |
    And no side effects
```

The behavior of Neo4j is exactly the same as the expected result, but in falkerDB, the result is different.

<img width="946" alt="image" src="https://github.com/FalkorDB/FalkorDB/assets/53161583/d2a5f271-871d-41de-bb57-4d7e8a8b5c6d">

**Comment by gkorland** [comment]
Thanks for reporting the issue, we already fixed this issue on an unstable branch and should include it in the next release.
You can verify the fix on this container: docker run -it --rm -p 6379:6379 -p 3000:3000 falkordb/falkordb:edge-rs

**Comment by DvirDukhan** [comment]
Verified against both engines and closing as fixed by the Rust migration.

**Reproduction**

```cypher
CREATE (a:A), (b:B)
CREATE (a)-[:T1]->(b),
       (b)-[:T2]->(a)

MATCH (a)
MERGE (b)
WITH *
OPTIONAL MATCH (a)--(b)
RETURN count(*)
```

| Engine | Result |
| --- | --- |
| `falkordb/falkordb:latest` (v4.20.1, C engine) | count(*) = 4 after MATCH/MERGE/WITH*/OPTIONAL MATCH — wrong, should be 6 as OPTIONAL MATCH should re-evaluate per row of the WITH-partitioned stream |
| `falkordb/falkordb:edge-rs` (Rust engine) | count(*) = 6 — matches expected TCK/Neo4j semantics |

TCK-derived scenario about query-plan partitioning affecting OPTIONAL MATCH cardinality after MERGE+WITH*. latest returns 4, rs returns 6 which matches the openCypher TCK expectation.

The Rust engine became the default development engine in #2279 and ships as the
rolling `edge-rs` Docker tag, so this issue is resolved there. Reopen if you still
hit it on `edge-rs`.
